trustfall_core/src/ir/ty.rs

```diff
@@ -0,0 +1,332 @@
+use core::fmt::{self, Formatter};
+use std::{fmt::Display, sync::Arc};
+
+use async_graphql_parser::types::{BaseType, Type as GQLType};
+use serde::{de::Visitor, Deserialize, Deserializer, Serialize, Serializer};
+
+/// A backing-storage independent immutable representation of a GraphQL type.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Type {
+    base: Arc<str>,
+    modifiers: Modifiers,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Modifiers {
+    mask: u64, // space for ~30 levels of list nesting
+}
+
+impl Modifiers {
+    const NON_NULLABLE_MASK: u64 = 1;
+    const LIST_MASK: u64 = 2;
+    const MAX_LIST_DEPTH: u64 = 30;
+    // represents the most far left list bit that can be set before adding a new list will overflow
+    const MAX_LIST_DEPTH_MASK: u64 = Self::LIST_MASK << ((Self::MAX_LIST_DEPTH - 1) * 2); // - 1 because we start shifted over once.
+
+    fn nullable(&self) -> bool {
+        (self.mask & Self::NON_NULLABLE_MASK) == 0
+    }
+
+    fn is_list(&self) -> bool {
+        (self.mask & Self::LIST_MASK) != 0
+    }
+
+    fn as_list(&self) -> Option<Modifiers> {
+        self.is_list().then_some(Modifiers { mask: self.mask >> 2 })
+    }
+
+    fn at_max_list_depth(&self) -> bool {
+        (self.mask & (Self::MAX_LIST_DEPTH_MASK)) == (Self::MAX_LIST_DEPTH_MASK)
```

Comment:
some of the parens are a bit excessive :)
```suggestion
        (self.mask & Self::MAX_LIST_DEPTH_MASK) == Self::MAX_LIST_DEPTH_MASK
```